In [1]:
import os
import sys
sys.path.append(os.chdir('../../'))

In [2]:
import pandas as pd
from tqdm.auto import tqdm
import numpy as np
import datetime

import matplotlib.pyplot as plt
import plotly.plotly as ply
import plotly.graph_objs as go
import src.utils.graph as graph
from plotly.offline import init_notebook_mode, iplot
init_notebook_mode(connected=True)

#from sklearn.multioutput import MultiOutputRegressor
from src.algorithms.multioutput import MultiOutputRegressor, RegressorChain

from sklearn.utils import shuffle
#from catboost import CatBoostRegressor
from src.algorithms.catboost_model import CatBoost

from sklearn.metrics import mean_absolute_error

tqdm.pandas()
pd.options.display.max_columns = None
pd.options.display.max_colwidth = -1

import src.data as data
import src.utility as utility

In [3]:
#dataset_train = base_df.merge(data.base_dataset(), how='left')
X, Y = data.dataset('train', onehot=False, drop_index_columns=True)

caching merged dataset train


In [3]:
#path = os.path.join(f'resources/dataset/preprocessed/base_dataframe_train_inferred.csv.gz')
#dataset = pd.read_csv(path, parse_dates=True, index_col=0)
#X, Y = data.dataset_interpolated('train', onehot=False)

In [5]:
X.head(10)

EMERGENCY_LANE  LANES  ROAD_TYPE  EVENT_DETAIL             EVENT_TYPE  \
1554  1               3      1          25            Calamita_naturale       
1370  1               3      1          25            Calamita_naturale       
1745  1               3      1          25            Calamita_naturale       
1639  1               3      1          25            Calamita_naturale       
1972  1               3      1          25            Calamita_naturale       
4963  0               1      0          55            Opera_arte              
3979  1               3      0          68            Gestione_viabilita      
4564  0               2      0          15            Pavimentazione          
448   0               1      0          34            Segnaletica_verticale   
5927  0               2      0          0             Gestione_viabilita      

      SPEED_AVG_-4  SPEED_AVG_-3  SPEED_AVG_-2  SPEED_AVG_-1  SPEED_SD_-4  \
1554  225.8283      228.5009      240.6095      222.6471      32.3762       
1370  232.1462      224.8530      237.0677      230.2363      33.9988       
1745  209.1308      208.2986      217.8098      210.9917      25.5567       
1639  189.9103      213.9138      225.2186      218.1234      46.2343       
1972  192.5420      194.8749      202.7207      209.9104      22.0654       
4963 NaN            169.6589      152.0920     NaN           NaN            
3979  195.1580      193.2793      193.8842      194.4184      23.8040       
4564  224.0723      227.7082      229.2097      231.3410      30.2731       
448   185.2610      180.8728      185.8003      188.0476      17.7785       
5927  204.8268      205.1152      195.2856      197.2301      30.5589       

      SPEED_SD_-3  SPEED_SD_-2  SPEED_SD_-1  SPEED_MAX_-4  SPEED_MAX_-3  \
1554  28.2049      35.5813      27.4555      343.016       309.038        
1370  43.0886      26.9673      29.9258      310.656       315.510        
1745  27.5619      36.3047      33.1063      273.442       284.768        
1639  21.7366      27.6684      24.5135      257.262       278.296        
1972  24.3183      26.6896      32.4509      241.082       257.262        
4963  30.8755      12.5747     NaN          NaN            213.576        
3979  22.1790      20.4265      22.4175      266.970       288.004        
4564  31.3228      33.6854      33.7112      300.948       312.274        
448   22.4662      17.9861      17.6903      218.430       234.610        
5927  36.4896      27.7364      28.4026      270.206       309.038        

      SPEED_MAX_-2  SPEED_MAX_-1  SPEED_MIN_-4  SPEED_MIN_-3  SPEED_MIN_-2  \
1554  375.376       304.184       135.912       181.216       182.834        
1370  339.780       347.870       74.428        84.136        177.980        
1745  351.106       362.432       160.182       147.238       148.856        
1639  309.038       292.858       69.574        160.182       129.440        
1972  275.060       334.926       142.384       135.912       105.170        
4963  166.654      NaN           NaN            134.294       131.058        
3979  263.734       275.060       147.238       150.474       140.766        
4564  396.410       318.746       166.654       147.238       165.036        
448   270.206       232.992       152.092       147.238       156.946        
5927  268.588       255.644       165.036       124.586       132.676        

      SPEED_MIN_-1  N_VEHICLES_-4  N_VEHICLES_-3  N_VEHICLES_-2  \
1554  176.362       131.0          107.0          113.0           
1370  177.980       88.0           66.0           79.0            
1745  139.148       95.0           107.0          120.0           
1639  155.328       83.0           91.0           92.0            
1972  134.294       104.0          86.0           110.0           
4963 NaN           NaN             7.0            6.0             
3979  137.530       321.0          316.0          352.0           
4564  156.946       78.0           64.0           77.0            


In [4]:
print(X.shape)
print(Y.shape)
print()
print(Y.dtypes)

(33202, 77)
(33202, 4)

SPEED_AVG_Y_0    float64
SPEED_AVG_Y_1    float64
SPEED_AVG_Y_2    float64
SPEED_AVG_Y_3    float64
dtype: object


## Preprocess data

In [4]:
X.head(10)

EMERGENCY_LANE  LANES  ROAD_TYPE  EVENT_DETAIL  \
0  1               2      1          11             
1  1               2      1          6              
2  1               2      1          6              
3  1               2      1          6              
4  1               2      1          2              
5  1               2      1          2              
6  1               2      1          2              
7  1               2      1          11             
8  1               2      1          2              
9  1               2      1          2              

                    EVENT_TYPE  SPEED_AVG_-4  SPEED_AVG_-3  SPEED_AVG_-2  \
0  Veicolo_in_avaria           NaN            199.6642      199.3150       
1  Gestione_viabilita          NaN           NaN           NaN             
2  Gestione_viabilita          NaN           NaN            208.9124       
3  Gestione_viabilita          NaN           NaN           NaN             
4  Manutenzione_opere_in_verde  210.7883      213.6824      223.9108       
5  Manutenzione_opere_in_verde  223.1689      218.5277      219.2802       
6  Manutenzione_opere_in_verde  218.5277      219.2802      214.6672       
7  Veicolo_in_avaria           NaN           NaN            199.6642       
8  Manutenzione_opere_in_verde NaN            216.4407      208.1131       
9  Manutenzione_opere_in_verde  211.3166      212.3555      207.6177       

   SPEED_AVG_-1  SPEED_SD_-4  SPEED_SD_-3  SPEED_SD_-2  SPEED_SD_-1  \
0  204.7759     NaN           41.4457      52.0643      43.0659       
1  208.9124     NaN          NaN          NaN           38.8807       
2  201.9654     NaN          NaN           38.8807      39.7029       
3  213.5957     NaN          NaN          NaN           35.7280       
4  223.1689      34.5211      33.1529      32.9331      30.7040       
5  214.6672      30.7040      29.5927      28.9895      32.1558       
6  216.8678      29.5927      28.9895      32.1558      29.6041       
7  199.3150     NaN          NaN           41.4457      52.0643       
8  217.6035     NaN           44.8391      42.4317      37.3565       
9  213.4945      27.7924      37.2444      34.9689      33.5346       

   SPEED_MAX_-4  SPEED_MAX_-3  SPEED_MAX_-2  SPEED_MAX_-1  SPEED_MIN_-4  \
0 NaN            263.734       292.858       294.476      NaN             
1 NaN           NaN           NaN            352.724      NaN             
2 NaN           NaN            352.724       286.386      NaN             
3 NaN           NaN           NaN            334.926      NaN             
4  281.532       343.016       389.938       341.398       59.866         
5  341.398       328.454       300.948       286.386       59.866         
6  328.454       300.948       286.386       310.656       59.866         
7 NaN           NaN            263.734       292.858      NaN             
8 NaN            368.904       281.532       349.488      NaN             
9  278.296       307.420       273.442       296.094       61.484         

   SPEED_MIN_-3  SPEED_MIN_-2  SPEED_MIN_-1  N_VEHICLES_-4  N_VEHICLES_-3  \
0  59.866        59.866        59.866       NaN             107.0           
1 NaN           NaN            59.866       NaN            NaN              
2 NaN            59.866        59.866       NaN            NaN              
3 NaN           NaN            59.866       NaN            NaN              
4  61.484        64.720        59.866        231.0          228.0           
5  59.866        61.484        59.866        239.0          265.0           
6  61.484        59.866        59.866        265.0          255.0           
7 NaN            59.866        59.866       NaN            NaN              
8  59.866        59.866        59.866       NaN             183.0           
9  59.866        59.866        59.866        280.0          232.0           

   N_VEHICLES_-2  N_VEHICLES_-1    WEATHER_-4    WEATHER_-3    WEATHER_-2  \
0  129.0          139.0          NaN           Qu

In [7]:
# add features
import src.preprocessing.other_features as feat
avg_speed_road_event = feat.avg_speed_for_roadtype_event()
X = X.merge(avg_speed_road_event, how='left', on=['EVENT_TYPE','ROAD_TYPE'])

avg_speed_road_event.head()

In [4]:
X.fillna(0, inplace=True)
#Y.fillna(0, inplace=True)

weather_cols = [col for col in X.columns if col.startswith('WEATHER_')]
categorical_cols = ['EMERGENCY_LANE', 'ROAD_TYPE', 'EVENT_DETAIL','EVENT_TYPE',] + weather_cols

## Train/Validation split

In [12]:
#X, Y = shuffle(X, Y)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, shuffle=False)

In [6]:
def evaluate(model, X_test, y_test):
    mask_test = np.all(y_test.notnull(), axis=1)
    print('number of valid samples:', (mask_test*1).sum())

    y_pred = model.predict(X_test[mask_test])
    return mean_absolute_error(y_test[mask_test], y_pred)

## Fit

In [17]:
from catboost import CatBoostRegressor

mae = []
for colY in Y.columns:
    mask = Y[colY].notnull()
    x = X_train[mask]
    y = y_train[colY][mask]
    
    mask_val = y_test[colY].notnull()
    x_val = X_test[mask]
    y_val = y_test[colY][mask_val]
    
    catboost = CatBoostRegressor(**{
        'loss_function': 'MAE',
        'eval_metric': 'MAE',
        'iterations':2000,
        'depth':6,
        'learning_rate':1,
        'early_stopping_rounds': 80,
        'cat_features': categorical_cols,
        'verbose': True
    })
    
    catboost.fit(x, y, eval_set=[(x_val, y_val)])
    
    # evaluate
    y_pred = catboost.predict(x_val)
    
    mae.append( mean_absolute_error(y_val, y_pred) )

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:6: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:10: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



0:	learn: 177.7051518	test: 183.3838661	best: 183.3838661 (0)	total: 64.2ms	remaining: 2m 8s
1:	learn: 177.2066703	test: 182.8855700	best: 182.8855700 (1)	total: 134ms	remaining: 2m 13s
2:	learn: 176.7080109	test: 182.3863849	best: 182.3863849 (2)	total: 209ms	remaining: 2m 19s
3:	learn: 176.2095406	test: 181.8884537	best: 181.8884537 (3)	total: 287ms	remaining: 2m 23s
4:	learn: 175.7110258	test: 181.3898264	best: 181.3898264 (4)	total: 360ms	remaining: 2m 23s
5:	learn: 175.2125470	test: 180.8924187	best: 180.8924187 (5)	total: 430ms	remaining: 2m 22s
6:	learn: 174.7142695	test: 180.3983990	best: 180.3983990 (6)	total: 511ms	remaining: 2m 25s
7:	learn: 174.2153597	test: 179.8998603	best: 179.8998603 (7)	total: 576ms	remaining: 2m 23s
8:	learn: 173.7174475	test: 179.4053919	best: 179.4053919 (8)	total: 651ms	remaining: 2m 23s
9:	learn: 173.2189454	test: 178.9068115	best: 178.9068115 (9)	total: 718ms	remaining: 2m 22s
10:	learn: 172.7208487	test: 178.4085969	best: 178.4085969 (10)	total:

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:6: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:10: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



0:	learn: 178.1178345	test: 183.6057419	best: 183.6057419 (0)	total: 57.3ms	remaining: 1m 54s
1:	learn: 177.6200324	test: 183.1065959	best: 183.1065959 (1)	total: 117ms	remaining: 1m 56s
2:	learn: 177.1205405	test: 182.6070087	best: 182.6070087 (2)	total: 155ms	remaining: 1m 43s
3:	learn: 176.6215992	test: 182.1077964	best: 182.1077964 (3)	total: 212ms	remaining: 1m 45s
4:	learn: 176.1230408	test: 181.6090495	best: 181.6090495 (4)	total: 277ms	remaining: 1m 50s
5:	learn: 175.6239486	test: 181.1108092	best: 181.1108092 (5)	total: 330ms	remaining: 1m 49s
6:	learn: 175.1243486	test: 180.6119179	best: 180.6119179 (6)	total: 365ms	remaining: 1m 44s
7:	learn: 174.6245850	test: 180.1121434	best: 180.1121434 (7)	total: 387ms	remaining: 1m 36s
8:	learn: 174.1263118	test: 179.6192451	best: 179.6192451 (8)	total: 438ms	remaining: 1m 36s
9:	learn: 173.6264302	test: 179.1193639	best: 179.1193639 (9)	total: 454ms	remaining: 1m 30s
10:	learn: 173.1274715	test: 178.6201117	best: 178.6201117 (10)	total

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:6: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:10: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



0:	learn: 178.2720632	test: 183.6882458	best: 183.6882458 (0)	total: 57.5ms	remaining: 1m 54s
1:	learn: 177.7745480	test: 183.1911266	best: 183.1911266 (1)	total: 116ms	remaining: 1m 55s
2:	learn: 177.2775558	test: 182.6960076	best: 182.6960076 (2)	total: 170ms	remaining: 1m 53s
3:	learn: 176.7798193	test: 182.2001034	best: 182.2001034 (3)	total: 223ms	remaining: 1m 51s
4:	learn: 176.2814143	test: 181.7015984	best: 181.7015984 (4)	total: 287ms	remaining: 1m 54s
5:	learn: 175.7825381	test: 181.2045834	best: 181.2045834 (5)	total: 341ms	remaining: 1m 53s
6:	learn: 175.2840116	test: 180.7057302	best: 180.7057302 (6)	total: 394ms	remaining: 1m 52s
7:	learn: 174.7852006	test: 180.2066894	best: 180.2066894 (7)	total: 446ms	remaining: 1m 51s
8:	learn: 174.2872730	test: 179.7097339	best: 179.7097339 (8)	total: 506ms	remaining: 1m 51s
9:	learn: 173.7882992	test: 179.2122012	best: 179.2122012 (9)	total: 564ms	remaining: 1m 52s
10:	learn: 173.2887831	test: 178.7132252	best: 178.7132252 (10)	total

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:6: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.

/Users/federico/miniconda3/envs/recsys/lib/python3.6/site-packages/ipykernel_launcher.py:10: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



0:	learn: 178.3600737	test: 183.8659000	best: 183.8659000 (0)	total: 54.2ms	remaining: 1m 48s
1:	learn: 177.8614294	test: 183.3669732	best: 183.3669732 (1)	total: 113ms	remaining: 1m 52s
2:	learn: 177.3632222	test: 182.8696069	best: 182.8696069 (2)	total: 167ms	remaining: 1m 51s
3:	learn: 176.8644544	test: 182.3708401	best: 182.3708401 (3)	total: 222ms	remaining: 1m 50s
4:	learn: 176.3670827	test: 181.8730577	best: 181.8730577 (4)	total: 279ms	remaining: 1m 51s
5:	learn: 175.8691310	test: 181.3810003	best: 181.3810003 (5)	total: 336ms	remaining: 1m 51s
6:	learn: 175.3704847	test: 180.8832939	best: 180.8832939 (6)	total: 389ms	remaining: 1m 50s
7:	learn: 174.8722230	test: 180.3840807	best: 180.3840807 (7)	total: 444ms	remaining: 1m 50s
8:	learn: 174.3749322	test: 179.8856283	best: 179.8856283 (8)	total: 504ms	remaining: 1m 51s
9:	learn: 173.8761420	test: 179.3869698	best: 179.3869698 (9)	total: 559ms	remaining: 1m 51s
10:	learn: 173.3777698	test: 178.8943586	best: 178.8943586 (10)	total

In [18]:
mae

[6.6273399128792825, 7.877166646134302, 9.378274779374266, 10.505643840577255]

In [22]:
sum(mae)/4

8.597106294741277

In [6]:
catboost = CatBoost(params_dict={
    'X': X,
    'loss_function': 'MAE',
    'eval_metric': 'MAE',
    'iterations':2,
    'depth':6,
    'learning_rate':1,
    'cat_features': categorical_cols,
    'verbose': False
})

In [8]:
model = RegressorChain(catboost)
model.fit(X, Y)

catboost: fitting...
catboost: fitting...
catboost: fitting...
catboost: fitting...


RegressorChain(base_estimator=<src.algorithms.catboost_model.CatBoost object at 0x1a20daf3c8>,
        cv=None, order=None, random_state=None)

In [11]:
evaluate(X_test, y_test)

7.8075137571574285

## Baseline

Catboost chain (no features + weather): 7.076924911258332 - TEST: 7.9718490219958404

Catboost chain: 7.8075137571574285

Catboost chain (avg speed x road type): 7.096708023412791

Catboost chain (avg speed x road type + avg speed sensor + avg_speed_road): 6.79054199721705

## Test model

In [3]:
import src.algorithms.inout as inout

In [4]:
X, Y = data.dataset('test', onehot=False, drop_index_columns=True)
X.fillna(0, inplace=True)
X.shape

caching merged dataset test


(2981, 78)

In [5]:
X.head()

EMERGENCY_LANE  LANES  ROAD_TYPE  EVENT_DETAIL               EVENT_TYPE  \
2749  0               1      0          8             Ostacolo_in_carreggiata   
1335  0               2      2          56            Meteo                     
1313  0               2      2          56            Meteo                     
1605  0               2      0          56            Meteo                     
1608  0               2      0          56            Meteo                     

      SPEED_AVG_-4  SPEED_AVG_-3  SPEED_AVG_-2  SPEED_AVG_-1  SPEED_SD_-4  \
2749  164.9204      162.1236      166.6540      157.5683      11.9988       
1335  196.1016      226.2889      192.1375      194.6223      20.1049       
1313  240.3352      225.6634      218.0255      284.7680      33.1693       
1605  270.4371      201.6028      0.0000        0.0000        37.0641       
1608  212.5648      0.0000        205.4860      0.0000        42.7506       

      SPEED_SD_-3  SPEED_SD_-2  SPEED_SD_-1  SPEED_MAX_-4  SPEED_MAX_-3  \
2749  13.2142      13.1212      17.4809      187.688       179.598        
1335  29.7050      28.4664      24.3725      218.430       273.442        
1313  35.2651      38.4380      7.0527       299.330       288.004        
1605  12.6070      0.0000       0.0000       330.072       220.048        
1608  0.0000       26.0895      0.0000       317.128       0.000          

      SPEED_MAX_-2  SPEED_MAX_-1  SPEED_MIN_-4  SPEED_MIN_-3  SPEED_MIN_-2  \
2749  187.688       195.778       152.092       152.092       144.002        
1335  228.138       232.992       165.036       190.924       155.328        
1313  307.420       292.858       177.980       160.182       173.126        
1605  0.000         0.000         210.340       186.070       0.000          
1608  242.700       0.000         177.980       0.000         171.508        

      SPEED_MIN_-1  N_VEHICLES_-4  N_VEHICLES_-3  N_VEHICLES_-2  \
2749  126.204       14.0           5.0            18.0            
1335  161.800       10.0           14.0           16.0            
1313  279.914       26.0           34.0           24.0            
1605  0.000         14.0           10.0           0.0             
1608  0.000         16.0           0.0            5.0             

      N_VEHICLES_-1    WEATHER_-4    WEATHER_-3    WEATHER_-2    WEATHER_-1  \
2749  26.0           Quasi Sereno  Quasi Sereno  Quasi Sereno  Quasi Sereno   
1335  7.0            Quasi Sereno  Quasi Sereno  Quasi Sereno  Quasi Sereno   
1313  3.0            Quasi Sereno  Quasi Sereno  Quasi Sereno  Quasi Sereno   
1605  0.0            Quasi Sereno  Quasi Sereno  0             0              
1608  0.0            Quasi Sereno  0             Quasi Sereno  0              

      DISTANCE_-4  DISTANCE_-3  DISTANCE_-2  DISTANCE_-1  TEMPERATURE_-4  \
2749  14.0         14.0         14.0         14.0         4.0              
1335  55.0         55.0         55.0         55.0         4.0              
1313  48.0         48.0         48.0         48.0         4.0              
1605  80.0         80.0         0.0          0.0          6.0              
1608  47.0         0.0          47.0         0.0          6.0              

      TEMPERATURE_-3  TEMPERATURE_-2  TEMPERATURE_-1  MIN_TEMPERATURE_-4  \
2749  4.0             4.0             4.0             1.0                  
1335  4.0             4.0             4.0             1.0                  
1313  4.0             4.0             4.0             1.0                  
1605  6.0             0.0             0.0             5.0                  
1608  0.0             5.0             0.0             5.0                  

      MIN_TEMPERATURE_-3  MIN_TEMPERATURE_-2  MIN_TEMPERATURE_-1  \
2749  1.0                 1.0                 1.0                  
1335  1.0                 1.0                 1.0                  
1313  1.0                 1.0                 1.0                  
1605  5.0                 0.0                 0.0                  
1608

In [6]:
#model = inout.load('saved_models/catboost_multiout_5.98269_baseline_5ft.jl')
model = inout.load('saved_models/catboost_multiout_8.45851_baseline.jl')

In [7]:
inout.evaluate(model, X_test=X, y_test=Y)

This evaluation will be done for samples having ALL y not null!


(8.352709228393282,
 [7.060148226263454, 8.086440020790906, 8.995492885807689, 9.268755780711079])